java.source.base/src/org/netbeans/modules/java/source/usages/BuildArtifactMapperImpl.java

```diff
@@ -871,6 +890,38 @@ private boolean isUpdateResources(@NullAllowed final File targetFolder) {
             return new File(targetFolder, TAG_UPDATE_RESOURCES).exists();
         }
     }
+
+    private static boolean targetNewerThanSourceFile(File target, URL approximateSource) {
+        if (!"file".equals(approximateSource.getProtocol())) {
+            LOG.log(Level.FINER, "#227791: ignoring non-file-based source {0}", approximateSource);
+            return false;
+        }
+        if (!target.isFile()) {
+            LOG.log(Level.FINER, "#227791: {0} does not even exist", target);
+            return false;
+        }
+        long targetLastMod = target.lastModified();
+        File mockSrc;
+        try {
+            mockSrc = BaseUtilities.toFile(approximateSource.toURI());
+        } catch (URISyntaxException x) {
+            LOG.log(Level.FINER, "#227791: cannot convert " + approximateSource, x);
+            return false;
+        }
+        File src = new File(mockSrc.getParentFile(), mockSrc.getName().replaceFirst("([$].+)*[.]sig$", ".java"));
```

Comment:
... what does this other thing you do with .sig here?

Reply:
Here we are going from, say, `/prj/src/main/java/pkg/Outer$Inner.sig` (obviously not a real path; constructed at line 847) to `/prj/src/main/java/pkg/Outer.java` which we suspect is an actual file whose timestamp we could compare to that of the corresponding `/userdir/var/cache/index/s1234/java/15/classes/pkg/Outer$Inner.sig`.

Reply:
So, I first thought this patch is only copying the sig files if they are newer than the classfiles.

But if I understand it correctly, this patch goes much farther. Consider these two sources:
---A.java
public class A {

    public static void main(String[] args) {
        B.hello(1);
    }
    
}
---B.java
public class B {
    
    public static void hello(int i) {
        System.err.println("Hello!");
    }

}
---

Run A (with CoS). Then change B.java to:
---B.java
public class B {
    
    public static void hello(long i) {
        System.err.println("Hello!");
    }

}
---

Save and run (A) again:
---
run:
Exception in thread "main" java.lang.NoSuchMethodError: B.hello(I)V
	at A.main(A.java:4)
/home/lahvac/src/nb/apache-netbeans.jdk10/nbbuild/testuserdir/var/cache/executor-snippets/run.xml:111: The following error occurred while executing this line:
/home/lahvac/src/nb/apache-netbeans.jdk10/nbbuild/testuserdir/var/cache/executor-snippets/run.xml:94: Java returned: 1
BUILD FAILED (total time: 0 seconds)
---

It (seems to) work OK without this patch. Is this what we wanted? In particular, did we want to force this unconditionally to everyone (who is using CoS)? I can understand some can live with this bug, but can all our users? Aren't we exchanging a set of (undisclosed) probably fixable bugs for (one) systematic unfixable known obvious bug?

Am I missing something?

Reply:
> I first thought this patch is only copying the sig files if they are newer than the classfiles.

No, it is skipping the `*.sig` file copy when the `*.class` is newer than the `*.java`. In other words, trust the build system over the IDE: only run IDE-generated bytecode corresponding to source files the user has actually edited since the last external build.

> Aren't we exchanging a set of (undisclosed) probably fixable bugs for (one) systematic unfixable known obvious bug?

Yes, that is one perspective. Another is that we are mitigating a set of bugs which have not in fact been fixed in the past decade or so and which had no viable workaround, for a bug with two simple workarounds: do a clean build¹; or simply make a no-op edit to `A.java` (SPACE BACKSPACE) before rerunning it.

My experience has been that the kind of change you describe (`int` → `long` without any modifications to the caller) is pretty unusual; I do not recall ever hitting such a problem in the time that I have been running with this patch, whereas the kinds of problems that were routinely breaking CoS execution and interrupting my work have disappeared. YMMV.

The question is whether this special case can also be handled without reducing the overall benefit of the patch. I am doubtful. The example you describe is that `A.class` has a newer timestamp than `A.java` but `A.sig` has a newer timestamp than either (because the IDE’s internal compiler does dependency tracking), and we would prefer to pick `A.sig` since it is calling the new method signature. However as a user you have no control over when the IDE’s internal compiler decides to update `*.sig` files, and many NetBeans users are familiar with the dreaded phenomenon of unexpected classpath scanning and bogus error badges. That is the problem: an IDE bug can spray unwanted `*.sig` file updates all over your classpath. You can hope that there have been no such updates since your last clean build, but this is pretty fragile. In fact clean builds can themselves sometimes trigger scanning, for example when you have generated source roots.

¹Or an incremental build, depending on how sophisticated your project’s build system is. For example, NB Ant-based projects work correctly in this case after you simply `compile`.

Reply:
So, another example (with this patch), but this time needs a code change to make it deterministic: add:
try {Thread.sleep(3000);} catch (Throwable t) {}
as the very first line of BuildArtifactsMapperImpl's
private Boolean performUpdate(@NonNull final Context ctx) throws IOException {
(this is to ensure timestamps change)

Then create a J2SE project with two files:
```
---cos/test/CosTest.java
package cos.test;

public class CosTest {

    public static void main(String[] args) {
//        System.err.println(Dep.get());
    }
    
}
---cos/test/Dep.java
package cos.test;

public class Dep {

//    static String get() {
//        return "Hello!";
//    }
    
}
---
```

-do a clean build
-(CoS) run
-uncomment the line in CosTest, wait 5+seconds
-go to Dep, uncomment the method in Dep
-wait a few seconds, (CoS) run:
```
run:
Exception in thread "main" java.lang.RuntimeException: Uncompilable source code - Erroneous tree type: <any>
	at cos.test.CosTest.main(CosTest.java:6)
/home/lahvac/src/nb/jdk-modules/nbbuild/testuserdir/var/cache/executor-snippets/run.xml:111: The following error occurred while executing this line:
/home/lahvac/src/nb/jdk-modules/nbbuild/testuserdir/var/cache/executor-snippets/run.xml:94: Java returned: 1
BUILD FAILED (total time: 0 seconds)
```
-clean the project (do not build)
-(CoS) run:
```
run:
Hello!
BUILD SUCCESSFUL (total time: 0 seconds)
```

I.e. does not seem to be an indexing problem, but rather a CoS copying problem.

Reply:
> I was also digging around compile-time constants

Yes, this is another of the corner cases where the bytecode produced by a source file can change depending on changes in _other_ source files. There are some others involving inheritance, imports, etc. Again my intuition is that these cases are not important in practice compared to overall robustness, but perhaps I have just been blessed to work on projects with large Java code bases written using “interesting” design patterns and too-clever build scripts? It would be easy enough to make this patch be enabled via yet another checkbox, at the cost of UI complexity.

> nor apisupport.ant projects use it

Huh, I thought we used to use it for NB modules?

> it would be useful if bugs in the indexing would get reported

Agreed. It can get tricky to produce reports with steps to reproduce, especially in proprietary sources. `var/log/*.dump` files _could_ help; I have thousands of them, all apparently “coupling errors”, but am not sure what to _do_ with them exactly. There are also various bug reports with `AssertionError`s from (patched) javac internals, which sometimes get fixed and sometimes not.

> CoS also supports source code that does not compile

FWIW I have never used this feature myself.

> this patch complicates that significantly.

Does it? If you have uncompilable sources, clearly you introduced these changes since the last external build—or your build would have failed. So those edits should be producing new `*.java` and `*.sig` files, which CoS execution would still honor.

> CoS can be disentangled from java.source, and can be done solely in the projects, avoiding any indexing issues altogether.

Not sure what you mean by this.

Reply:
>> nor apisupport.ant projects use it

> Huh, I thought we used to use it for NB modules?

So, how do I enable it?

>> CoS also supports source code that does not compile

> FWIW I have never used this feature myself.

Me neither. If we can generalize that, I think it opens paths to very significant simplifications.

>> this patch complicates that significantly.

> Does it? If you have uncompilable sources, clearly you introduced these changes since the last external build—or your build would have failed. So those edits should be producing new *.java and *.sig files, which CoS execution would still honor.

See above for a particular example.

>> CoS can be disentangled from java.source, and can be done solely in the projects, avoiding any indexing issues altogether.

> Not sure what you mean by this.

Well, if ability to run erroneous source is not needed, then there's no strong reason why CoS in projects would need to use the outcomes of indexing. So the project (Maven/Ant/...) could simply rebuild what it sees fit, by means it sees fit (e.g. using ToolsProvider, or running a simplified version of the build script, etc.). Yes, this would mean some of the sources would be compiled twice, but:
a) indexing could, I think, be significantly simplified if it didn't have to produce runnable classfiles, improving robustness and performance.
b) CoS would not depend on the un-robust (dreaded) indexing, presumably improving quality for CoS users

Reply:
> -uncomment the line in CosTest, wait 5+seconds
> -go to Dep, uncomment the method in Dep

Are you saving the sources here, or letting the save be done by F6? If the latter, that does sound like a race condition here: you would want to let all newly saved sources be scanned into sig files before the run starts, which the IDE normally ensures, and perhaps that is getting broken by this change? Or if the former (manual saves in reverse order), then, well, do not do that. Either way, you can simply touch `CosTest` again to work around.

> the project (Maven/Ant/...) could simply rebuild what it sees fit, by means it sees fit (e.g. using ToolsProvider, or running a simplified version of the build script, etc.).

Ah, I get your point now. Using an external build script is not going to work well, since this adds launch overhead, and there could be many different modified sources with different build roots that need recompilation, which would in general require multiple tool executions in sequence. Also there is typically a difference between the classpath entries used by an external build (e.g., `~/.m2/repository/upstream/1.0-SNAPSHOT/upstream-1.0-SNAPSHOT.jar`) vs. the IDE’s notion (e.g., `/src/upstream/target/classes`).

So you could use `javax.tools.JavaCompiler`. Compared to using nbjavac, this is safer, since the only IDE bugs which could break CoS would involve a failure to determine the correct classpath and other compilation options—i.e., the same limitations as CoS generally. You still do need to topologically sort the compilations.

> indexing could […] be significantly simplified if it didn't have to produce runnable classfiles

Do I remember correctly that at one time `*.sig` files merely captured non-`private` member signatures, so were deliberately incomplete bytecode, and that this was changed to generating complete class files at some point? I also vaguely recall some design constraint imposed by the need to support unpacked WAR files.